test/e2e/vmagent_test.go

```diff
@@ -291,3 +302,411 @@ func createExternalNodeCRD(data *TestData, nodeName string, ifName string, ip st
+		if vm.osType == osType {
+			return vm, vmList[(i+1)%len(vmList)], nil
+		}
+	}
+	return vmInfo{}, vmInfo{}, fmt.Errorf("not found a VM configured with OS type %s in vmList", osType)
+}
+
+func addLabelsOnExternalNode(t *testing.T, data *TestData, namespace string, vmList []vmInfo) {
+	for _, vm := range vmList {
+		nodeName := vm.nodeName
+		en, err := data.crdClient.CrdV1alpha1().ExternalNodes(namespace).Get(context.TODO(), nodeName, metav1.GetOptions{})
+		assert.Nil(t, err, fmt.Sprintf("Failed to get ExternalNode %s: %v", nodeName, err))
+		if en.Labels == nil {
+			en.Labels = make(map[string]string)
+		}
+		en.Labels[externalNodeLabelKey] = nodeName
+		_, err = data.crdClient.CrdV1alpha1().ExternalNodes(namespace).Update(context.TODO(), en, metav1.UpdateOptions{})
+		assert.Nil(t, err, fmt.Sprintf("Failed to add label on ExternalNode %s: %v", nodeName, err))
+	}
+}
+
+func testANPWithFQDN(t *testing.T, data *TestData, name string, namespace string, appliedToVM vmInfo, allowedURLs []string, droppedURLs []string, rejectedURLs []string) error {
+	var err error
+	allURLs := append(append(allowedURLs, droppedURLs...), rejectedURLs...)
+	for _, url := range allURLs {
+		err := runCurlCommandOnVM(t, data, appliedToVM, url, crdv1alpha1.RuleActionAllow)
+		assert.Nil(t, err)
+	}
+
+	fqdnSettings := make(map[string]*crdv1alpha1.RuleAction, 0)
+	for _, url := range allowedURLs {
+		action := crdv1alpha1.RuleActionAllow
+		fqdnSettings[url] = &action
+	}
+	for _, url := range droppedURLs {
+		action := crdv1alpha1.RuleActionDrop
+		fqdnSettings[url] = &action
+	}
+	for _, url := range rejectedURLs {
+		action := crdv1alpha1.RuleActionReject
+		fqdnSettings[url] = &action
+	}
+
+	anp := createANPWithFQDN(t, data, name, namespace, appliedToVM, fqdnSettings)
+	for url, action := range fqdnSettings {
+		err = runCurlCommandOnVM(t, data, appliedToVM, url, *action)
+		assert.Nil(t, err)
+	}
+	err = data.DeleteANP(anp.Namespace, anp.Name)
+	require.Nil(t, err)
+	for _, url := range allURLs {
+		err := runCurlCommandOnVM(t, data, appliedToVM, url, crdv1alpha1.RuleActionAllow)
+		assert.Nil(t, err)
+	}
+	return nil
+}
+
+func testANPWithConnectivity(t *testing.T, data *TestData, name string, namespace string, appliedToVM vmInfo, peerVM vmInfo, proto AntreaPolicyProtocol, ingress bool, ruleAction crdv1alpha1.RuleAction, matchPeerEntity bool) error {
+	if proto != ProtocolTCP && proto != ProtocolUDP && proto != ProtocolICMP {
+		return fmt.Errorf("unsupported protocol %s when applying ANP rule to ExternalNode", proto)
+	}
+
+	var srcVM, dstVM vmInfo
+	if ingress {
+		srcVM = peerVM
+		dstVM = appliedToVM
+	} else {
+		srcVM = appliedToVM
+		dstVM = peerVM
+	}
+
+	var initialConnected, anpConnected bool
+	if ruleAction == crdv1alpha1.RuleActionDrop || ruleAction == crdv1alpha1.RuleActionReject {
+		initialConnected = true
+		anpConnected = false
+	} else if ruleAction == crdv1alpha1.RuleActionAllow {
+		initialConnected = false
+		anpConnected = true
```

Comment:
`testANPWithConnectivity` assumes the initial state could be default drop when the policy's action is Allow, but `testANPWithFQDN` assumes the initial state is always default allow, why the difference?